channelserver/src/meta.rs

```diff
@@ -331,23 +414,79 @@ mod test {
+        blank_headers.insert(
+            http::header::USER_AGENT,
+            http::header::HeaderValue::from_static(blank_header),
+        );
         assert_eq!(None, get_ua(&blank_headers, None));
         let empty_headers = http::HeaderMap::new();
         assert_eq!(None, get_ua(&empty_headers, None));
     }
 
-    #[ignore]
     #[test]
-    fn test_location() {
+    fn test_location_good() {
         let test_ip = "63.245.208.195"; // Mozilla
 
         let langs = vec!["en".to_owned()];
         let mut sender = SenderData::default();
         sender.remote = Some(test_ip.to_owned());
         // TODO: either mock maxminddb::Reader or pass it in as a wrapped impl
+        let iploc = maxminddb::Reader::open("mmdb/latest/GeoLite2-City.mmdb").unwrap();
+        get_location(&mut sender, &langs, None, &iploc);
+        assert_eq!(sender.city, Some("Sacramento".to_owned()));
+        assert_eq!(sender.region, Some("California".to_owned()));
+        assert_eq!(sender.country, Some("United States".to_owned()));
+    }
+
+    #[test]
+    fn test_location_bad() {
+        let test_ip = "192.168.1.1";
+
+        let langs = vec!["en".to_owned()];
+        let mut sender = SenderData::default();
+        sender.remote = Some(test_ip.to_owned());
+        // TODO: either mock maxminddb::Reader or pass it in as a wrapped impl
+        let iploc = maxminddb::Reader::open("mmdb/latest/GeoLite2-City.mmdb").unwrap();
+        get_location(&mut sender, &langs, None, &iploc);
+        assert_eq!(sender.city, None);
+        assert_eq!(sender.region, None);
+        assert_eq!(sender.country, None);
+    }
+
+    #[test]
+    fn test_get_remote() {
+        let mut headers = actix_web::http::header::HeaderMap::new();
+
+        let proxy_list:Vec<IpNet> = vec!("192.168.0.0/24".parse().unwrap());
+
+        headers.insert(
+            http::header::HeaderName::from_lowercase("x-forwarded-for".as_bytes()).unwrap(),
+            "10.10.10.10, 192.168.0.1".parse().unwrap(),
+        );
+
+        let remote = get_remote(&headers, &proxy_list);
+        assert_eq!(remote, Some("10.10.10.10".to_owned()));
+
+        // Adding a header which should override the previous "success"
+        headers.insert(
+            http::header::HeaderName::from_lowercase("x-forwarded-for".as_bytes()).unwrap(),
+            "10.11.11.11, 192.168.0.1".parse().unwrap(),
+        );
+
+        let remote = get_remote(&headers, &proxy_list);
+        assert_eq!(remote, Some("10.11.11.11".to_owned()));
+
+        // Adding the Primary header
+        headers.insert(http::header::HeaderName::from_lowercase("forwarded".as_bytes()).unwrap(),
+            "by=10.10.10.10;proto=http;for=10.12.12.12;host=10.13.13.13,for=192.168.0.1;by=10.09.09.09".parse().unwrap());
+
+        let remote = get_remote(&headers, &proxy_list);
+        assert_eq!(remote, Some("10.12.12.12".to_owned()));
```

Comment:
I tried to think of a suite of testcases to communicate expected behavior here.  I'm very happy to be convinced that these are not correct, but I'd like to understand why.

These all assume that `get_remote()` takes `req.peer_addr()` as first argument.

With no X-Forwarded-For, we should take the peer address:
* `get_remote("1.2.3.4", {}, ["192.168.0.0/24"]) == "1.2.3.4"`

When the peer address is not a trusted proxy, we should ignore X-Forwarded-For, since it's untrusted data:
* `get_remote("1.2.3.4", {"X-F-F":"2.3.4.5, 3.4.5.6"}, ["192.168.0.0/24"]) == "1.2.3.4"`

When the peer address is a trusted proxy, we should take the rightmost X-F-F entry that is not a trusted proxy:
* `get_remote("192.168.0.4", {"X-F-F":"1.2.3.4"}, ["192.168.0.0/24"]) == "1.2.3.4"`
* `get_remote("192.168.0.4", {"X-F-F":"1.2.3.4, 2.3.4.5"}, ["192.168.0.0/24"]) == "2.3.4.5"`
* `get_remote("192.168.0.4", {"X-F-F":"1.2.3.4, 2.3.4.5, 129.168.0.9"}, ["192.168.0.0/24"]) == "2.3.4.5"`

When the peer address is a trusted proxy, we should insist that X-F-F contain at least one entry, otherwise the trusted proxy is not behaving as expected:
* `get_remote("192.168.0.4", {}, ["192.168.0.0/24"]) => error!`
* `get_remote("192.168.0.4", {"X-F-F": ""}, ["192.168.0.0/24"]) => error!`
* `get_remote("192.168.0.4", {"X-F-F": "unparseable, garbage, here"}, ["192.168.0.0/24"]) => error!`

Do these make sense?

Reply:
Yep. going to resort those mostly because of additive reasons, but will add those tests.